chore: migrate trial Infobox [DET-3014 DET-3016]

hamidzr opened the pull request:
## Description

- infobox 
- add/compelete decoders for step, validation, checkpoint

## Test Plan

![Screen Capture_select-area_20200730131600](https://user-images.githubusercontent.com/12127420/88970187-d3322f00-d266-11ea-9424-45325bc683af.png)

## Commentary (optional)

- hyperparameter data blocked by https://github.com/determined-ai/determined/pull/971
- I did not include step to batch conversion to keep the PR smaller and gather more wrt the remove steps project.


<!-- User-facing API changes need the "User-facing API Change" label -->

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")

## Description

Lead with the intended commit body in this description field. For breaking changes, please include "BREAKING CHANGE:" at the beginning of your commit body. At minimum, this section should include a bracketed reference to the Jira ticket, e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.

## Test Plan

Describe the situations in which you've tested your change, and/or a screenshot as appropriate. Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.

## Commentary (optional)

Use this section of your description to add context to the PR. Could be for particularly tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc. You may intentionally leave this section blank and remove the title.
--->



[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

hkang1 commented on webui/react/src/types.ts:
```diff
@@ -211,25 +211,25 @@ export enum CheckpointState {
 }
 
-export interface LatestValidationMetrics {
+export interface ValidationMetrics {
```
👍

hkang1 commented on webui/react/src/components/TupleRow.tsx:
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+
+import css from './TupleRow.module.scss';
+
+interface Props {
+  label: string;
+  children?: React.ReactNode;
+}
+
+const TupleRow: React.FC<Props> = ({ label, children }: Props) => {
+
+  if (children === undefined) return <></>;
+  return (
+    <tr className={css.base} key={label}>
+      <td className={css.label}>{label}</td>
+      <td>
+        {[ 'string', 'number' ].includes(typeof children) ?
+          <span>{children}</span> : children
+        }
+      </td>
+    </tr>
+  );
+};
+
+export default TupleRow;
```
**suggestion:** The `TupleRow` doesn't quite stand on its own as a component just because it can't render on it's own and requires `table` and `tbody` to render. Instead, this is a good opportunity to build a `InfoBox` component with a separate `Row` subcomponent within that component.

hamidzr replied:
good point. what do you think is a good interface for this infobox component? how about a list of tuples?

hkang1 replied:
How about...
```
{ label: string, info?: React.ReactNode }[];
```
And if the `info` field is not specified, it will skip rendering it.

hkang1 commented:
**suggestion:** check out [PR 974](https://github.com/determined-ai/determined/pull/974) for info box setup and style tweaks.

hkang1 commented on webui/react/src/ioTypes.ts:
```diff
@@ -21,6 +21,8 @@ export const decode = <T>(type: io.Mixed, data: any): T => {
   }
 };
 
+const ioNullUndefined = io.union([ io.null, io.undefined ]);
```
nice, this is great

hamidzr replied:
should save us from some surprises.

hamidzr replied:
I was gonna go ahead and update ioTypes to use this in more places but to avoid more conflicts and delaying this work I'll leave that for another PR

hkang1 replied:
sounds great

hkang1 commented on webui/react/src/types.ts:
```diff
@@ -211,25 +211,25 @@ export enum CheckpointState {
+  trialId: number;
+}
+
+export interface Checkpoint extends BaseStep {
```
**suggestion (simplify):** I could just be me, but I found the multiple levels of extensions to be much harder to follow. One benefit of this approach is that you can isolate changes to properties of `BaseStep` in a single area, but it also puts us in a tougher place if we want different properties for different interfaces. My recommendation is to just keep the `BaseStep` properties in the respective interfaces.

hamidzr replied:
we would have less levels if we could extend multiple things.. actually we might be able to with `&` :thinking:

hamidzr replied:
I moved the types closer together (in code location) so they're more easily accessible.

>  but it also puts us in a tougher place if we want different properties for different interfaces

I'm not be missing something from your POV but I think if those properties are truly shared as they are defined here then we shouldn't have this case? :thinking: 

BaseStep now has 4 properties which would be copied across 3 (so far) types if we remove it. I'd lean toward the DRY principle unless we have specific reasons not to : ) let's zoom if you wanna chat more about this.

hamidzr replied:
we discussed this and we reduced the nesting by extending multiple types.

hkang1 commented on webui/react/src/types.ts:
```diff
@@ -251,11 +251,16 @@ export interface TrialItem {
+  metrics?: ValidationMetrics;
+}
+
+export interface Step extends BaseStep {
   state: RunState;
+  checkpoint?: Checkpoint;
+  validation?: Validation;
```
**question:** can these properties just be part of `BaseStep`?

hamidzr replied:
as it's set now no? that would give the impression as if checkpoint or validation had sub checkpoints.

hamidzr commented:
I created an infobox component and used that.

hkang1 commented on webui/react/src/ioTypes.ts:
```diff
@@ -141,55 +143,69 @@ const checkpointStates: Record<string, null> = Object.values(CheckpointState)
   id: io.number,
-  latest_validation_metrics: io.union([ ioLatestValidatonMetrics, io.null ]),
+  latest_validation_metrics: io.union([ ioValidationMetrics, io.null ]),
   num_batches: io.union([ io.number, io.null ]),
```
@hamidzr  **issue:** I just encountered this, could you apply the `ioNullOrUndefined` here on `num_batches` as well if you haven't already? 🙏

hamidzr commented:
> LGTM! nice work with this PR as usual

thanks I moved the pages around and deleted the changes I had made to experiment info box with the file moved I'll just leave it as is and we change it in another pr

hkang1 commented on webui/react/src/pages/TrialDetails/TrialInfoBox.tsx:
```diff
@@ -10,7 +10,8 @@ import { humanReadableBytes, humanReadableFloat } from 'utils/string';
 import { checkpointSize, trialDurations } from 'utils/types';
 
-import CheckpointModal from './CheckpointModal';
+import CheckpointModal from '../../components/CheckpointModal';
```
**chore:** this can just be `components/CheckpointModal`

hkang1 commented on webui/react/src/pages/TrialDetails/TrialInfoBox.tsx:
```diff
@@ -93,7 +93,8 @@ const TrialInfoBox: React.FC<Props> = ({ trial, experiment }: Props) => {
     },
     {
-      info: `${humanReadableFloat(bestValidation)} (${experiment.config.searcher.metric})`,
+      info: bestValidation && `${humanReadableFloat(bestValidation)}` +
+      `(${experiment.config.searcher.metric})`,
```
**nit:** I think this strips out a space between the readable float and the metric type in parentheses. How about:
```
info: bestValidation && `
  ${humanReadableFloat(bestValidation)} (${experiment.config.searcher.metric})
`,
```

hamidzr replied:
IMO this looks better 
```
      info: bestValidation && `${humanReadableFloat(bestValidation)} ` +
      `(${experiment.config.searcher.metric})`,
```
but either works

hamidzr replied:
or the same one you have with the ` on the same line
